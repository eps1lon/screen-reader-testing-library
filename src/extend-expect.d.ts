export {};
export { extendExpect as default } from "./index";

declare global {
	namespace jest {
		interface Matchers<R> {
			toAnnounceNVDA(expectedLines: string[][]): Promise<void>;
			toMatchSpeechSnapshot(snapshotName?: string): Promise<void>;
			toMatchSpeechInlineSnapshot(expectedLinesSnapshot?: string): void;
		}
	}
}
