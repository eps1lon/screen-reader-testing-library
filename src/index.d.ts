export type SpeechLine = string[];
export type Speech = string[][];
export type Recorder = {
    start: () => Promise<void>;
    /**
     * Returns the speech output since the last `start()`.
     */
    stop: () => Promise<string[][]>;
    /**
     * Returns the speech output during the call to `fn`.
     */
    record: (fn: () => Promise<void>) => Promise<string[][]>;
    readable: () => Promise<boolean>;
};
export function awaitNvdaRecording(): Promise<void>;
/**
 * @param {string} logFilePath
 * @returns {Recorder}
 */
export function createSpeechRecorder(logFilePath: string): Recorder;
/**
 * @param {string} logFilePath
 */
export function createJestSpeechRecorder(logFilePath: string): Recorder;
/**
 * @param {jest.Expect} expect
 * @param {string} logFilePath
 * @returns {void}
 */
export function extendExpect(expect: jest.Expect, logFilePath: string): void;
