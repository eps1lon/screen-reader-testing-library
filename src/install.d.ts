export type ScreenReaderDescriptor = {
    name: string;
    revision: string;
};
