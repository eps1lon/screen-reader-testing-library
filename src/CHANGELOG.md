# screen-reader-testing-library

## 0.0.6

### Patch Changes

- [`730096e`](https://github.com/eps1lon/screen-reader-testing-library/commit/730096e2f641a8bba167adae583b99130fc14673) Thanks [@eps1lon](https://github.com/eps1lon)! - Fix failing download on install

  Changesets use a different naming scheme for tags which we forgot to incorporate in the install script.

## 0.0.5

### Patch Changes

- [`a71cb81`](https://github.com/eps1lon/screen-reader-testing-library/commit/a71cb8163caaf0cdc9ce96509e3483e94b35cdac) [#16](https://github.com/eps1lon/screen-reader-testing-library/pull/16) Thanks [@eps1lon](https://github.com/eps1lon)! - Implement same signature as `toMatchSpeechSnapshot` in `toMatchSpeechInlineSnapshot`

## 0.0.4

### Patch Changes

- [`8947f5b`](https://github.com/eps1lon/screen-reader-testing-library/commit/8947f5b34a2f5511143efa5d0cea6b429413fc16) [#13](https://github.com/eps1lon/screen-reader-testing-library/pull/13) Thanks [@eps1lon](https://github.com/eps1lon)! - Add TypeScript types
