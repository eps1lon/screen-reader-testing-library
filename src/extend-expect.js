const { default: diff } = require("jest-diff");
const { toMatchInlineSnapshot, toMatchSnapshot } = require("jest-snapshot");
const { createSpeechRecorder } = require("./index");

const speechSnapshotBrand = Symbol.for(
	"screen-reader-testing-library.speechSnapshot"
);

/**
 * Must return `any` or `expect.extend(createMatchers(logFilePath))` does not typecheck.
 * `toMatchInlineSnapshot` will be unassignable for unknown reasons.
 * @param {string} logFilePath
 * @returns {any}
 */
function createMatchers(logFilePath) {
	const speechRecorder = createSpeechRecorder(logFilePath);

	/**
	 * @param {import('./index').Speech} recordedSpeech
	 * @param {string} [expectedSpeechSnapshot]
	 * @returns {ReturnType<typeof toMatchInlineSnapshot>}
	 * @this {import('jest-snapshot/build/types').Context}
	 */
	function toMatchSpeechInlineSnapshot(recordedSpeech, expectedSpeechSnapshot) {
		// Abort test on first mismatch.
		// Subsequent actions will be based on an incorrect state otherwise and almost always fail as well.
		this.dontThrow = () => {};
		if (typeof recordedSpeech === "function") {
			throw new Error(
				"Recording lines is not implemented by the matcher. Use `expect(recordLines(async () => {})).resolves.toMatchInlineSnapshot()` instead"
			);
		}

		const actualSpeechSnapshot = {
			[speechSnapshotBrand]: true,
			speech: recordedSpeech,
		};

		// jest's `toMatchInlineSnapshot` relies on arity.
		if (expectedSpeechSnapshot === undefined) {
			return toMatchInlineSnapshot.call(this, actualSpeechSnapshot);
		}
		return toMatchInlineSnapshot.call(
			this,
			actualSpeechSnapshot,
			expectedSpeechSnapshot
		);
	}

	/**
	 *
	 * @param {() => Promise<void>} fn
	 * @param {string} [snapshotName]
	 * @returns {Promise<ReturnType<typeof toMatchSnapshot>>}
	 * @this {import('jest-snapshot/build/types').Context}
	 */
	async function toMatchSpeechSnapshot(fn, snapshotName) {
		const speech = await speechRecorder.record(fn);

		return toMatchSnapshot.call(this, speech, snapshotName);
	}

	/**
	 * @param {() => Promise<void>} fn
	 * @param {import('./index').Speech} expectedSpeech
	 * @returns {Promise<{actual: unknown, message: () => string, pass: boolean}>}
	 * @this {import('jest-snapshot/build/types').Context}
	 */
	async function toAnnounceNVDA(fn, expectedSpeech) {
		const actualSpeech = await speechRecorder.record(fn);

		const options = {
			comment: "deep equality",
			isNot: this.isNot,
			promise: this.promise,
		};

		const pass = this.equals(actualSpeech, expectedSpeech);
		const message = pass
			? () =>
					this.utils.matcherHint("toBe", undefined, undefined, options) +
					"\n\n" +
					`Expected: not ${this.utils.printExpected(expectedSpeech)}\n` +
					`Received: ${this.utils.printReceived(actualSpeech)}`
			: () => {
					const diffString = diff(expectedSpeech, actualSpeech, {
						expand: this.expand,
					});
					return (
						this.utils.matcherHint("toBe", undefined, undefined, options) +
						"\n\n" +
						(diffString && diffString.includes("- Expect")
							? `Difference:\n\n${diffString}`
							: `Expected: ${this.utils.printExpected(expectedSpeech)}\n` +
							  `Received: ${this.utils.printReceived(actualSpeech)}`)
					);
			  };

		return { actual: actualSpeech, message, pass };
	}

	return { toAnnounceNVDA, toMatchSpeechSnapshot, toMatchSpeechInlineSnapshot };
}
// This module is only a convenience helper augmenting jest types.
// TypeScript users should import from this

/**
 * @param {jest.Expect} expect
 * @param {string} logFilePath
 * @returns {void}
 */
export default function extendExpect(expect, logFilePath) {
	expect.extend(createMatchers(logFilePath));

	expect.addSnapshotSerializer({
		/**
		 * @param {any} val
		 */
		print(val) {
			/**
			 * @type {{ speech: import('./index').Speech }}
			 */
			const snapshot = val;
			const { speech } = snapshot;

			return speech
				.map((line) => {
					return `"${line.join(", ")}"`;
				})
				.join("\n");
		},
		test(value) {
			return value != null && value[speechSnapshotBrand] === true;
		},
	});
}
