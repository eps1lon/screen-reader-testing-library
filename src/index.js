const { promises: fs } = require("fs");
const { default: diff } = require("jest-diff");
const { toMatchInlineSnapshot, toMatchSnapshot } = require("jest-snapshot");
const { extractSpeechLines } = require("./logParser");

/**
 * @param {number} timeoutMS
 * @returns {Promise<void>}
 */
function sleep(timeoutMS) {
	return new Promise((resolve) => setTimeout(() => resolve(), timeoutMS));
}

// need to wait for NVDA to flush the speech
// It seems like it debounces the speech so that the user isn't flooded when quickly navigating
function awaitNvdaRecording() {
	return sleep(2000);
}

/**
 * @param {string} logFilePath
 */
function createSpeechRecorder(logFilePath) {
	// Intuitively we keep an open handle and read new logs every time we call `readAllNewLogs`.
	// However, once node read streams have reached the current! end calling read() again won't read anything that was newly written.
	// We have to refresh the file.
	// It's a bit heave on memory since we read the whole file into memory.
	// Should be fine for most tests.
	let logFileOffset = 0;
	async function stop() {
		await awaitNvdaRecording();

		const fullContent = await fs.readFile(logFilePath, { encoding: "utf-8" });
		const newContent = fullContent.slice(logFileOffset);
		logFileOffset = fullContent.length;

		return extractSpeechLines(newContent);
	}

	async function start() {
		const logFile = await fs.stat(logFilePath);
		logFileOffset = logFile.size;
	}

	/**
	 * @param {() => Promise<void>} fn
	 * @returns {Promise<string[][]>}
	 */
	async function recordLines(fn) {
		// move to end
		await start();
		await fn();
		return await stop();
	}

	function readable() {
		return fs.access(logFilePath);
	}

	return { readable, recordLines, start, stop };
}

/**
 * @param {string} logFilePath
 */
function createMatchers(logFilePath) {
	const recorder = createSpeechRecorder(logFilePath);

	/**
	 *
	 * @param {() => Promise<void>} fn
	 * @param {string[][]} _expectedLines
	 * @returns {Promise<ReturnType<typeof toMatchInlineSnapshot>>}
	 * @this {import('jest-snapshot/build/types').Context}
	 */
	async function toMatchSpeechInlineSnapshot(fn, _expectedLines) {
		// throws with "Jest: Multiple inline snapshots for the same call are not supported."
		throw new Error("Not implemented");
		// // move to end
		// await recorder.start();
		// await fn();
		// const actualLines = await recorder.stop();

		// return toMatchInlineSnapshot.call(this, actualLines);
	}

	/**
	 *
	 * @param {() => Promise<void>} fn
	 * @param {string} [snapshotName]
	 * @returns {Promise<ReturnType<typeof toMatchSnapshot>>}
	 * @this {import('jest-snapshot/build/types').Context}
	 */
	async function toMatchSpeechSnapshot(fn, snapshotName) {
		const actualLines = await recorder.recordLines(fn);

		return toMatchSnapshot.call(this, actualLines, snapshotName);
	}

	/**
	 * @param {() => Promise<void>} fn
	 * @param {string[][]} expectedLines
	 * @returns {Promise<{actual: unknown, message: () => string, pass: boolean}>}
	 * @this {import('jest-snapshot/build/types').Context}
	 */
	async function toAnnounceNVDA(fn, expectedLines) {
		const actualLines = await recorder.recordLines(fn);

		const options = {
			comment: "deep equality",
			isNot: this.isNot,
			promise: this.promise,
		};

		const pass = this.equals(actualLines, expectedLines);
		const message = pass
			? () =>
					this.utils.matcherHint("toBe", undefined, undefined, options) +
					"\n\n" +
					`Expected: not ${this.utils.printExpected(expectedLines)}\n` +
					`Received: ${this.utils.printReceived(actualLines)}`
			: () => {
					const diffString = diff(expectedLines, actualLines, {
						expand: this.expand,
					});
					return (
						this.utils.matcherHint("toBe", undefined, undefined, options) +
						"\n\n" +
						(diffString && diffString.includes("- Expect")
							? `Difference:\n\n${diffString}`
							: `Expected: ${this.utils.printExpected(expectedLines)}\n` +
							  `Received: ${this.utils.printReceived(actualLines)}`)
					);
			  };

		return { actual: actualLines, message, pass };
	}

	return { toAnnounceNVDA, toMatchSpeechInlineSnapshot, toMatchSpeechSnapshot };
}

/**
 * @param {string} logFilePath
 */
function createJestSpeechRecorder(logFilePath) {
	const recorder = createSpeechRecorder(logFilePath);

	beforeAll(async () => {
		try {
			await recorder.readable();
		} catch (error) {
			throw new Error(`Log file in '${logFilePath}' is not readable`);
		}
	});

	return recorder;
}

module.exports = {
	awaitNvdaRecording,
	createSpeechRecorder,
	createMatchers,
	createJestSpeechRecorder,
};
