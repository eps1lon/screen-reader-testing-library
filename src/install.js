const crypto = require("crypto");
const extract = require("extract-zip");
const fs = require("fs-extra");
const http = require("http");
const https = require("https");
const path = require("path");
const lockfile = require("proper-lockfile");
const os = require("os");
const URL = require("url");

const dev = process.env.SRTL_DEV;

installScreenReaders().catch((error) => {
	console.error(error);
	process.exit(1);
});

/**
 * @typedef {object} ScreenReaderDescriptor
 * @property {string} name
 * @property {string} revision
 */

/**
 * Installs screen readers that can be used with this library.
 *
 * Based on https://github.com/microsoft/playwright/blob/cdd9fd6b2e9e1c4fc3c4fce696fdfcf30a13c222/src/install/installer.ts
 */
async function installScreenReaders() {
	const packagePath = __dirname;
	const screenReadersPath = getScreenReadersPath();

	await fs.mkdirp(screenReadersPath);
	const lockfilePath = path.join(screenReadersPath, "__dirlock");
	const releaseLock = await lockfile.lock(screenReadersPath, {
		retries: {
			retries: 10,
			// Retry 20 times during 10 minutes with
			// exponential back-off.
			// See documentation at: https://www.npmjs.com/package/retry#retrytimeoutsoptions
			factor: 1.27579,
		},
		onCompromised:
			/**
			 * @param {Error} error
			 */
			(error) => {
				throw new Error(`${error.message} Path: ${lockfilePath}`);
			},
		lockfilePath,
	});

	const linksDir = path.join(screenReadersPath, ".links");
	await fs.mkdirp(linksDir);
	await fs.writeFile(path.join(linksDir, sha1(packagePath)), packagePath);

	await validateCache(packagePath, screenReadersPath, linksDir);
	await releaseLock();
}

/**
 * @param {string} packagePath
 * @returns {Promise<ScreenReaderDescriptor[]>}
 */
async function readScreenReadersToDownload(packagePath) {
	const screenReaders = JSON.parse(
		(await fs.readFile(path.join(packagePath, "screenReaders.json"))).toString()
	)["screenReaders"];
	return screenReaders;
}

/**
 * @param {string} packagePath
 * @param {string} screenReadersPath
 * @param {string} linksDir
 * @returns {Promise<void>}
 */
async function validateCache(packagePath, screenReadersPath, linksDir) {
	// 1. Collect used downloads and package descriptors.
	/**
	 * @type {Set<string>}
	 */
	const usedScreenReaderPaths = new Set();
	for (const fileName of await fs.readdir(linksDir)) {
		const linkPath = path.join(linksDir, fileName);
		let linkTarget = "";
		try {
			linkTarget = (await fs.readFile(linkPath)).toString();
			const screenReadersToDownload = await readScreenReadersToDownload(
				linkTarget
			);
			for (const screenReader of screenReadersToDownload) {
				const usedScreenReaderPath = getScreenReaderDirectory(
					screenReadersPath,
					screenReader
				);
				const screenReaderRevision = parseInt(screenReader.revision, 10);
				const markerFileExists = await fs
					.access(
						getScreenReaderMarkerFilePath(screenReadersPath, screenReader)
					)
					.then(
						() => true,
						() => false
					);
				if (markerFileExists) {
					usedScreenReaderPaths.add(usedScreenReaderPath);
				}
			}
		} catch (e) {
			await fs.unlink(linkPath).catch((e) => {});
		}
	}
	// 2. Delete all unused browsers.
	let downloadedScreenReaders = (await fs.readdir(screenReadersPath)).map(
		(file) => {
			return path.join(screenReadersPath, file);
		}
	);
	downloadedScreenReaders = downloadedScreenReaders.filter((file) => {
		return isScreenReaderDirectory(file);
	});
	/**
	 * @type {Set<string>}
	 */
	const directories = new Set(downloadedScreenReaders);
	for (const screenReaderPath of usedScreenReaderPaths) {
		directories.delete(screenReaderPath);
	}
	for (const directory of directories) {
		await fs.remove(directory).catch((e) => {});
	}

	// 3. Install missing screen readers for this package.
	const myScreenReadersToDownload = await readScreenReadersToDownload(
		packagePath
	);
	const packageVersion = await fs
		.readFile(path.join(packagePath, "package.json"), { encoding: "utf8" })
		.then((packageJson) => {
			return JSON.parse(packageJson)["version"];
		});
	const artifactsUrlBase = `https://github.com/eps1lon/screen-reader-testing-library/releases/download/v${packageVersion}/`;
	for (const screenReader of myScreenReadersToDownload) {
		await downloadScreenReader(
			artifactsUrlBase,
			screenReadersPath,
			screenReader
		);
		await fs.writeFile(
			getScreenReaderMarkerFilePath(screenReadersPath, screenReader),
			""
		);
	}
}

/**
 * @param {string} data
 * @returns {string}
 */
function sha1(data) {
	const sum = crypto.createHash("sha1");
	sum.update(data);
	return sum.digest("hex");
}

// source: https://github.com/microsoft/playwright/blob/cdd9fd6b2e9e1c4fc3c4fce696fdfcf30a13c222/src/utils/browserPaths.ts#L122-L132
function getCacheDirectory() {
	if (process.platform === "linux")
		return process.env.XDG_CACHE_HOME || path.join(os.homedir(), ".cache");

	if (process.platform === "darwin")
		return path.join(os.homedir(), "Library", "Caches");

	if (process.platform === "win32")
		return (
			process.env.LOCALAPPDATA || path.join(os.homedir(), "AppData", "Local")
		);
	throw new Error("Unsupported platform: " + process.platform);
}

function getPackageCacheDirectory() {
	return path.join(getCacheDirectory(), "screen-reader-testing-library");
}

function getScreenReadersPath() {
	return path.join(getPackageCacheDirectory(), "screen-readers");
}

/**
 * @param {string} screenReadersPath
 * @param {ScreenReaderDescriptor} screenReader
 */
function getScreenReaderDirectory(screenReadersPath, screenReader) {
	return path.join(
		screenReadersPath,
		// screenreadersPath must contain a flat structure
		// we can't nest them like e.g. nvda/2020.3
		`${screenReader.name}-${screenReader.revision}`
	);
}

/**
 * @param {string} screenReaderPath
 * @returns {boolean}
 */
function isScreenReaderDirectory(screenReaderPath) {
	const baseName = path.basename(screenReaderPath);
	return baseName.startsWith("nvda-");
}

/**
 * @param {string} screenReadersPath
 * @param {ScreenReaderDescriptor} screenReader
 */
function getScreenReaderMarkerFilePath(screenReadersPath, screenReader) {
	return path.join(
		getScreenReaderDirectory(screenReadersPath, screenReader),
		"INSTALLATION_COMPLETE"
	);
}

/**
 * @param {string} artifactsUrlBase
 * @param {string} screenReadersPath
 * @param {ScreenReaderDescriptor} screenReader
 * @returns {Promise<void>}
 */
async function downloadScreenReader(
	artifactsUrlBase,
	screenReadersPath,
	screenReader
) {
	const screenReaderPath = getScreenReaderDirectory(
		screenReadersPath,
		screenReader
	);
	const screenReaderExists = await fs.access(screenReaderPath).then(
		() => true,
		() => false
	);
	if (screenReaderExists) {
		return;
	}

	const url = revisionURL(artifactsUrlBase, screenReader);
	const zipPath = path.join(
		os.tmpdir(),
		`screen-reader-testing-library-download-${screenReader.name}-${screenReader.revision}.zip`
	);

	try {
		if (dev) {
			await fs.copy(
				getScreenReaderDirectory(
					path.resolve(__dirname, "../vendor"),
					screenReader
				),
				screenReaderPath,
				{ recursive: true }
			);
		} else {
			await downloadFile(url, zipPath);
			await extract(zipPath, { dir: screenReaderPath });
		}
		// TODO: Do we need to do this?
		// await fs.chmod(browserPaths.executablePath(screenReaderPath, browser)!, 0o755);
	} catch (e) {
		process.exitCode = 1;
		throw e;
	} finally {
		await fs.unlink(zipPath).catch(() => {});
	}
}

/**
 *
 * @param {string} artifactsUrlBase
 * @param {ScreenReaderDescriptor} screenReader
 */
function revisionURL(artifactsUrlBase, screenReader) {
	return new URL.URL(
		`${screenReader.name}-${screenReader.revision}.zip`,
		artifactsUrlBase
	);
}

/**
 * @param {URL} url
 * @param {string} destinationPath
 */
function downloadFile(url, destinationPath) {
	let fulfill = () => {};
	/**
	 * @type {(error: Error) => void}
	 */
	let reject = () => {};

	const promise = new Promise((x, y) => {
		fulfill = x;
		reject = y;
	});

	const request = httpRequest(url, (response) => {
		if (response.statusCode !== 200) {
			const error = new Error(
				`Download failed: server returned code ${response.statusCode}. URL: ${url}`
			);
			// consume response data to free up memory
			response.resume();
			reject(error);
			return;
		}
		const file = fs.createWriteStream(destinationPath);
		file.on("finish", () => fulfill());
		file.on("error", (error) => reject(error));
		response.pipe(file);
	});
	request.on("error", (error) => reject(error));
	return promise;
}

/**
 * @param {URL|string} urlOrString
 * @param {(response: import('http').IncomingMessage) => void} callback+
 */
function httpRequest(urlOrString, callback) {
	const url =
		typeof urlOrString === "string" ? new URL.URL(urlOrString) : urlOrString;
	const requestCallback =
		/**
		 * @param {import('http').IncomingMessage} response
		 */
		(response) => {
			console.log(response.statusCode, urlOrString);
			if (
				response.statusCode !== undefined &&
				response.statusCode >= 300 &&
				response.statusCode < 400 &&
				response.headers.location
			) {
				httpRequest(response.headers.location, callback);
			} else {
				callback(response);
			}
		};
	const request =
		url.protocol === "https:"
			? https.request(url, requestCallback)
			: http.request(url, requestCallback);
	request.end();
	return request;
}
