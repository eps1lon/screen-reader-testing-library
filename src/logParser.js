/**
 * @param {string} nvdaLog
 * @returns {string[][]}
 */
function extractSpeechLines(nvdaLog) {
	return nvdaLog
		.split(/\r?\n/)
		.filter((line) => {
			return line.startsWith("Speaking ");
		})
		.map((line) => {
			// In: "Speaking ['speech1', 'speech2', 'speech 3']"
			// Out: "['speech1', 'speech2', 'speech 3']"
			const listText = line.trim().replace(/^Speaking \[([^\]]+)\]/, "$1");

			// In: "['speech1', 'speech2', 'speech 3']"
			// Out: the corresponding array structure in JS with added quotes
			return listText
				.split(",")
				.map((rawSpeech) => {
					const speech = rawSpeech.trim();
					if (speech.startsWith("'")) {
						return speech.slice(1, -1);
					}
					// ignore commands for now
					return null;
				})
				.filter(
					/**
					 * @returns {speech is string}
					 */
					(speech) => speech !== null
				);
		});
}

module.exports = { extractSpeechLines };
