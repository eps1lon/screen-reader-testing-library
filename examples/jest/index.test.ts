import * as playwright from "playwright";
import { awaitNvdaRecording, createMatchers } from "../../src";

const logFilePath = process.env.LOG_FILE_PATH;

declare global {
	namespace jest {
		interface Matchers<R> {
			toAnnounceNVDA(expectedLines: string[][]): Promise<void>;
			toMatchSpeechSnapshot(snapshotName?: string): Promise<void>;
			// throws with "Jest: Multiple inline snapshots for the same call are not supported."
			// toMatchSpeechInlineSnapshot(expectedLines: string[][]): R;
		}
	}
}

describe("chromium", () => {
	let browser: playwright.Browser;
	let page: playwright.Page;
	beforeAll(async () => {
		if (logFilePath === undefined) {
			throw new TypeError(
				"Log filepath not specified. Set the path in an environment variable named `LOG_FILE_PATH`."
			);
		}

		expect.extend({ ...(await createMatchers(logFilePath)) });
		browser = await playwright.chromium.launch({ headless: false });
		page = await browser.newPage();
	});

	afterAll(async () => {
		await browser.close();
	});

	beforeEach(async () => {
		page = await browser.newPage();
	});

	afterEach(async () => {
		// Closing the page results in different behavior for subsequent tests.
		// For some reason NVDA is focusing the language change dropdown when reloading the page.
		// Not closing pages and opening new ones works though.
		// await page.close();
	});

	it("produces the expected NVDA speech output when searching the docs", async () => {
		await page.goto("https://material-ui.com/");
		// Without bringing it to front the adress bar will still be focused.
		// NVDA wouldn't record any page actions
		await page.bringToFront();
		await awaitNvdaRecording();

		await expect(async () => {
			await page.keyboard.press("s");
		}).toAnnounceNVDA([
			["banner landmark"],
			[
				"Search",
				"combo box",
				"expanded",
				"has auto complete",
				"editable",
				"Searchâ€¦",
			],
		]);

		await expect(async () => {
			await page.keyboard.type("Rating");
		}).toAnnounceNVDA([]);

		await expect(async () => {
			await page.keyboard.press("ArrowDown");
		}).toAnnounceNVDA([["list"], ["Link to the result", "1 of 5"]]);
	}, 20000);

	it("matches the NVDA speech snapshot when searching the docs", async () => {
		await page.goto("https://material-ui.com/");
		// Without bringing it to front the adress bar will still be focused.
		// NVDA wouldn't record any page actions
		await page.bringToFront();
		await awaitNvdaRecording();

		await expect(async () => {
			await page.keyboard.press("s");
		}).toMatchSpeechSnapshot("Focused searchbox");

		await expect(async () => {
			await page.keyboard.type("Rating");
		}).toMatchSpeechSnapshot("entered searchterm");

		await expect(async () => {
			await page.keyboard.press("ArrowDown");
		}).toMatchSpeechSnapshot("navigated to first result");
	}, 20000);
});
