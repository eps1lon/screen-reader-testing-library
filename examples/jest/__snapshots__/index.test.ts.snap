// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chromium matches the NVDA speech snapshot when searching the docs: toMatchSpeechSnapshot 1`] = `
Array [
  Array [
    "banner landmark",
  ],
  Array [
    "Search",
    "combo box",
    "expanded",
    "has auto complete",
    "editable",
    "Searchâ€¦",
  ],
]
`;

exports[`chromium matches the NVDA speech snapshot when searching the docs: toMatchSpeechSnapshot 2`] = `Array []`;

exports[`chromium matches the NVDA speech snapshot when searching the docs: toMatchSpeechSnapshot 3`] = `
Array [
  Array [
    "list",
  ],
  Array [
    "Link to the result",
    "1 of 5",
  ],
]
`;
